src/System.Private.ServiceModel/tests/Common/Infrastructure/src/BridgeClientCertificateManager.cs

```diff
@@ -0,0 +1,394 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Net;
+using System.Security.Cryptography.X509Certificates;
+using System.Text;
+
+namespace Infrastructure.Common
+{
+    // This class manages the adding and removal of certificates.
+    // It also handles informing http.sys of which certificate to associate with SSL ports.
+    public static class BridgeClientCertificateManager
+    {
+        private static object s_certificateLock = new object();
+
+        // Resource names 
+        private const string CertificateAuthorityResourceName = "WcfService.CertificateResources.CertificateAuthorityResource";
+        private const string MachineCertificateResourceName = "WcfService.CertificateResources.MachineCertificateResource";
+
+        // key names in request/response keyval pairs
+        private const string EndpointResourceRequestNameResourceString = "name";
+        private const string SubjectKeyResourceString = "subject";
+        private const string ThumbprintKeyResourceString = "thumbprint";
+        private const string CertificateKeyResourceString = "certificate";
+        private const string IsLocalKeyResourceString = "isLocal";
+
+        // Keyed by the thumbprint of the certificate
+        private static string s_LocalFqdn = string.Empty;
+
+        // Dictionary of certificates installed by CertificateManager
+        // Keyed by the Thumbprint of the certificate
+        private static Dictionary<string, CertificateCacheEntry> s_myCertificates = new Dictionary<string, CertificateCacheEntry>(StringComparer.OrdinalIgnoreCase);
+        private static Dictionary<string, CertificateCacheEntry> s_rootCertificates = new Dictionary<string, CertificateCacheEntry>(StringComparer.OrdinalIgnoreCase);
+
+        private static string LocalFqdn
+        {
+            get
+            {
+                // Get an FQDN for the local machine and cache it
+                if (string.IsNullOrEmpty(s_LocalFqdn))
+                {
+                    s_LocalFqdn = Dns.GetHostEntryAsync("127.0.0.1").GetAwaiter().GetResult().HostName;
+                }
+                return s_LocalFqdn;
+            }
+        }
+
+        // Installs the root certificate from the bridge 
+        // Needed for all tests so we trust the incoming certificate coming from the service/bridge
+        // We do our best to detect if we're running on the same box as the bridge; if so, we don't try to install the cert
+        // as that operation requires admin privileges
+        public static void InstallRootCertificateFromBridge()
+        {
+            // PUT the Authority to the Bridge (returns thumbprint)
+            var response = BridgeClient.MakeResourcePutRequest(CertificateAuthorityResourceName, null);
```

Comment:
Can this method go re-entrant, and if so is that a problem making multiple PUT requests to this resource?